Add documentation to FF.com for SharedMap (#6867)
microsoft/FluidFramework: docs/content/docs/data-structures/map.md

const map = fluidContainer.initialObjects.customMap;
```

Finally, if you'd like to dynamically create `SharedMap` instances as part of the application lifecycle (i.e. if there are user interactions in the applications that require a new DDS to be created at runtime), you can add the `SharedMap` type to the `dynamicObjectTypes` field in the schema and call the container's `create` function.

```javascript
const schema = {
    /*...*/,
    dynamicObjectTypes: [ SharedMap ]
}

const { fluidContainer, containerServices } = await client.getContainer(/*service config*/, schema);

const newMap = await container.create(SharedMap); // Create a new SharedMap
```

Once the async call to `create` returns, you can treat it the same as you were using the `SharedMap` instances from your initial objects above. The only caveat here is that you will need to maintain a pointer to your newly created object. To store it in another `SharedMap`, please see the [Storing shared objects]({{< relref "#storing-shared-objects" >}}) section below and for general guidance on storing DDS references as handles, please see [here]({{< relref "dds.md#creating-and-storing-distributed-data-structures" >}})

> ## API Functionality

Reviewer: ```suggestion
## API
```

The `SharedMap` object provides a number of functions to allow you to edit the key/value pairs stored on the object. As stated earlier, these are intended to match the `Map` API. However, the keys used in `SharedMap` must be strings. Each edit will also trigger a `valueChanged` event which will be discussed in the [Events]({{< relref "#events" >}}) section below.

> - `set(key, value)` - Used for updating the value stored at `key` with the new provided value
> - `get(key)` - Returns the latest value stored on the key or `undefined` if it does not exist
> - `has(key)` - Checks to see if the key is available in the SharedMap and returns a boolean
> - `keys()` - Returns an iterator for all the keys that have been set for this map
> - `entries()` - Returns an iterator for all key/value pairs stored on the map
> - `delete(key)` - Removes the key/value from the map
> - `forEach(callbackFn: (value, key, map) => void)` - Applies the provided function to each entry in the map. For example, the following will print out all of the key/value pairs in the map

Reviewer: ```suggestion
- `set(key, value)` -- Used for updating the value stored at `key` with the new provided value
- `get(key)` -- Returns the latest value stored on the key or `undefined` if it does not exist
- `has(key)` -- Checks to see if the key is available in the SharedMap and returns a boolean
- `keys()` -- Returns an iterator for all the keys that have been set for this map
- `entries()` -- Returns an iterator for all key/value pairs stored on the map
- `delete(key)` -- Removes the key/value from the map
- `forEach(callbackFn: (value, key, map) => void)` -- Applies the provided function to each entry in the map. For example, the following will print out all of the key/value pairs in the map
```

```javascript
this.map.forEach((value, key) => console.log(`${key}-${value}`));

> ```

Reviewer: I think you'll need to indent this code block or it won't render under the `<ul><li>` in the output HTML. Maybe check a local build to test.

> - `clear()` - Removes all data from the map, deleting all of the keys and values stored within it

Reviewer: ```suggestion
- `clear()` -- Removes all data from the map, deleting all of the keys and values stored within it
```

---
title: SharedMap
menuPosition: 4
---
## Introduction

> The `SharedMap` distributed data structure (DDS) is designed to be a collaborative version of the [Map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) built-in object that is provided in JavaScript. It uses the same API, but ensures that any edits being made to the object are simultaneously transmitted to all clients. For example, in a traditional `Map`, setting a key would only set it on the local object. In a `SharedMap`, the moment one client sets that key, that update is automatically sent to all of the other connected clients who will update their local state with the new remote change. From a development standpoint, this allows you to develop against the `SharedMap` DDS as you would with a traditional `Map`, while ensuring that the data being updated is synced between all clients with all of the logic for managing and merging the changes abstracted away.

Author: These are now included in the starting "Tips" section to distinguish between Map and SharedMap

For additional background on DDSes and a general overview of their design, please take a look [here]({{< relref "dds.md" >}}).

## Creation

There are two paths to create an instance of the `SharedMap`:
1.  Using the [FluidContainer]({{< relref "containers.md" >}}) for when you'd like to directly interface with the `SharedMap` in your application

> 2. Using the DDS API directly for when you'd like to use it within your own custom data object

Author: Removed the DDS API section from here to instead focus on the consumption model patterns with `FluidContainer`

```javascript
const newMapKey = "uniqueMapId";
this.root.set(newMapKey, newMap.handle);
```
And then, you can load it back in the following manner:
```javascript
const newMapHandle = this.root.get(newMapKey);
const newMapLoaded = await newMapHandle.get();
```
`newMapLoaded` now points to the same DDS as your initial `newMap`

## API Functionality

The `SharedMap` object provides a number of functions to allow you to edit the key/value pairs stored on the object. As stated earlier, these are intended to match the `Map` API. However, the keys used in `SharedMap` must be strings. Each edit will also trigger a `valueChanged` event which will be discussed in the **Events** section below.

- `set(key, value)` - Used for updating the value stored at `key` with the new provided value
- `get(key)` - Returns the latest value stored on the key or `undefined` if it does not exist
- `has(key)` - Checks to see if the key is available. This can be used as a boolean check prior to using `get` to avoid returning an `undefined` value

> - `keys()` - Returns all the keys that have been set for this map

Author: Iterator, added to the description as well

- `entries()` - Returns an iterator for all values stored on the map. This can be easily converted into an array by doing `Array.from(map.entries())`
- `delete(key)` - Removes the key from the map
{{< callout note >}}

When removing a key from the `SharedMap`, please use the `delete` function instead of using `set` with `value` as `undefined`. The latter will not remove the key from the object and it will still be returned when developers call `keys` or `entries`

> {{< /callout >}}

Author: Committed in the earlier suggestion to include tips in the introduction to distinguish between Map and SharedMap. Also, I think developers may have their own reasons for storing an undefined value in the SharedMap (i.e. perhaps the  key is set but the value is still being loaded due to an async API call and its undefined until the async call finishes), similar to how they could store it into a Map. I was just trying to specify that they should call `delete` if they want to remove the value entirely, but perhaps that's unnecessary since that behavior would be the same as using Map itself

- `forEach(callbackFn)` - Applies the provided function to each entry in the map. For example, the following will print out all of the key/value pairs in the map
```javascript
this.map.forEach((value, key) => console.log(`${key}-${value}`));
```

> 

Author: Added in the signatures for both the selected API functions listed here and for the events below

- `clear()` - Removes all data from the map, deleting all of the keys and values stored within it

## Events

The `SharedMap` object will emit events any time a value on it changes, whether that be from an update applied to it by your local client or from a remote one. Specifically, there are two events to track here, 
- `valueChanged` - Fired anytime the map is modified due to a key being added, updated, or removed

> - `clear` - Fired when `clear()` is called to alert clients that all data from the map has been removed

Author: Edited above suggestion to use "sent" and committed

> 

Author: Added in the signatures for the events and their listeners, and moved the description for the parameters up to here so that its all in one place

button.textContent = "Randomize!";
const label = document.createElement('label');
    
button.addEventListener('click', () =>
    // Set the new value on the SharedMap
    map.set(dataKey, Math.random())
);

// Set up a callback to get the current value of the shared data whenever it changes and update the view.
const updateLabel = () => {
    const value = map.get(dataKey) || 0;
    label.textContent = `${value}`;
};
updateLabel();
// Use the valueChanged event to trigger the rerender whenever the value changes.
map.on('valueChanged', updateLabel);
```

In the code above, whenever a user clicks the button, it sets a new random value on your map's `dataKey`. This will result in a `valueChanged` event being fired for all of the clients who have this container open. Since `updateLabel` is a callback set up to update the view anytime this event gets fired, the view will always refresh with the new value for all users whenever any of the users clicks on the button.

The `valueChanged` event listener can also take in as parameters:
- a `changed` object of type `IValueChanged` which provides the `key` that was updated and what the `previousValue` was

> - a `local` boolean that indicates if the current client was the one that initiated the change

Author: Good point! Added the signatures to the events above and moved the description of the parameters up there as well

button.textContent = "Randomize!";
const label = document.createElement('label');
    
button.addEventListener('click', () =>
    map.set(dataKey, Math.random())
);

// Get the current value of the shared data to update the view whenever it changes.
const updateLabel = (changed, local) => {
    const value = map.get(dataKey) || 0;
    label.textContent = `${value} from ${local ? "me" : "someone else"}`;
    label.style.color = changed?.previousValue > value ? "red" : "green";
};
updateLabel(undefined, false);
    // Use the changed event to trigger the rerender whenever the value changes.
map.on('valueChanged', updateLabel);
```
Now, with the changes in `updateLabel`, the view will appropriately re-render to say if the value was last updated by the current user or by someone else. It will also compare the current value to the last one, and if the value has increased, it will set the text color to green. Otherwise, it will be red.

## Usage guides

`SharedMap` supports storing
- primitives (strings, numbers, booleans, etc.)
- [objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) 
- shared objects (DDSes, data objects)

> Let's walk through each of these different types and look at best practices on how to use them with the `SharedMap` APIs.

Author: Converted to level 2 headings and I can see them on the right nav pane now!

### Storing primitives

As you saw in the examples above, storing and fetching primitve values from a `SharedMap` are as simple as calling the `set`/`get` functions on the same key. The common paradigm to follow is:
1. Create a listener callback for updating the app state using `get` to fetch the latest data. This gets triggered on a `valueChanged` event
2. Adding some app functionality, whether that is through user action or otherwise, to call `set` with new data
3. Observe the `set` call trigger the `valueChanged` event for local and remote clients and see all of their local app states get updated

> Whichever client was the last to fire a `set` call will have its value be the current value on the key.

Author: nit: Made minor edit to above suggestion to fix numbering

### Storing objects

When storing objects in a `SharedMap`, the concepts for applying and fetching updates remain the same as when storing primitives. However, one thing to note is that all updates in `SharedMap` are merged using a last writer wins (LWW) strategy. This means that if multiple clients are writing values to the same key, whoever made the last update will "win" and overwrite the others. While this is fine for primitives, please be careful when using this to store objects if you are looking for different fields within the object to be modularly modified.

{{< callout warning >}}

If your application requires different fields in an object to be edited simultaneously by multiple people, do NOT store the entire object under one key. Instead, store each field under its own separate key such that each one of them can be updated separately.

> {{< /callout >}}

Author: I think the third tip in that section already covers this point now. However, having an additional warning here might be good to reinforce this concept for the developer since this section is where they'll actively be reading about how to store objects, and they may have missed the tip at the beginning

This would look something like this:
```json
{
    "task1": {
        "title": "Awesome Task",
        "description": "Doing the most awesome things",
        "assignedTo": {
            "name": "Joe Schmo",
            "email": "joeschmo@email.com",
            "address": "1234 Fluid Way"
        }
    }
}
```
Now lets consider the scenario where two users begin to edit the task at the same time. Let's have User A begin editing the `title` whereas User B is editing the `description`.

Each time that User A makes an edit, `map.set("task1", editedTask)` will get called where `editedTask` will hold the new `title` values. Similarly, whenever User B makes an edit, `map.set("task1", editedTask)` will also be called but `description` will now hold the updated values. However, since `SharedMap` uses the LWW merge strategy, whichever user sent their `editedTask` last will overwrite the others edit. If User A's `set` occurs last, `task1` will now hold the updates to the `title` but will have overwritten User B's updates to `description`. Similarly, if User B's `set` occurs last, `description` will be updated but not `title`.

There are two strategies to avoid this clash:

> 1. We can store each of these values in their own key and only hold the key at which they are stored in the `task1` object itself. This would mean your `SharedMap` could have an object like:

Author: Converted to level 3 headings

Now, you may be asking if you would need to do the same for `assignedTo` object. The answer is "yes" if the fields in this object were also something you would be expecting users modularly edit. However, if this object will always be set together then you can keep it under the same key. Since this user metadata is coming as a singular blob from a service, you can use the latter in this scenario.

```javascript
// Pseudocode for a function that makes a backend call to fetch a roster of all users in an array
const users = await getAvailableUsers();

// Let's say that the user the task is assigned to is at index pickedUserIndex
const pickedUser = users[pickedUserIndex];

// Now store this user object as a whole into the SharedMap
const task = map.get("task1");
map.set(task.assignedToKey, pickedUser)
```

> This will work as expected **because the entire object is being stored each time** instead of specific fields.
> 
> One way to think about this is that each value stored into the `SharedMap` is a unit of collaboration. Any time you want users to be able to collaborate on invidual pieces of an object, it should be appropriately stored in separate keys.

Reviewer: ```suggestion
This will work as expected **because the entire object is being stored each time** instead of specific fields.

One way to think about this is that each value stored into the `SharedMap` is the smallest simultaneously editable piece of data. Any time you want users to be able to simultaneously edit individual pieces of an object, you should store those properties in separate keys.
```

### Storing shared objects

> One of the powerful features of DDSes is that they are infintely nestable. A DDS can be stored in another DDS allowing you to dynamically set up your data hierarchy as best fits your application needs.

Reviewer: ```suggestion
One of the powerful features of DDSes is that they are nestable. A DDS can be stored in another DDS allowing you to dynamically set up your data hierarchy as best fits your application needs.
```

> This is done using the concept of [handles]({{< relref "data-modeling.md#using-handles-to-store-and-retrieve-fluid-objects" >}}). Handles provide a way for you to serialize your data object into a form that it can be stored into another DDS and loaded back from it.
> 
> To do this with `SharedMap`, all you need to do is specify an initial map as part of the `initialObjects` in the `ContainerSchema` and add the `SharedMap` type to `dynamicObjectTypes`.

Reviewer: ```suggestion
When storing a DDS within another DDS, you must store its [handle]({{< relref "data-modeling.md#using-handles-to-store-and-retrieve-fluid-objects" >}}), not the DDS itself. Similarly, when retrieving DDSes nested within other DDSes, you need to first get the object’s handle then get the object from the handle. This reference based approach allows the Fluid Framework to virtualize the data underneath, only loading objects when they are requested.

That’s all you need to know about handles in order to use DDSes effectively. If you want to learn more about handles, see [Fluid handles]({{< relref "handles.md" >}}).

The following example demonstrates nesting DDSes using `SharedMap`. You specify an initial SharedMap as part of the `initialObjects` in the `ContainerSchema` and add the `SharedMap` type to `dynamicObjectTypes`.
```

```javascript
const schema = {
    name: "example-container",
    initialObjects: {
        initalMap: SharedMap,
    },
    dynamicObjectTypes: [SharedMap]
}
```

Now, you can dynamically create additional `SharedMap` instances and store their handles into the initial map that is always provided in the container.

```javascript

> const schema = {
>     name: "example-container",
>     initialObjects: {
>         initalMap: SharedMap,
>     },
>     dynamicObjectTypes: [SharedMap]
> }
> 

Reviewer: ```suggestion
```
I think you can remove this since it's right above.

const { fluidContainer, containerServices } = await client.getContainer(/*service config*/, schema);

const initialMap = fluidContainer.initialObjects.initialMap;

> const newSharedMap = fluidContainer.create(SharedMap);
> 
> initialMap.set("newSharedMapKey", newSharedMap.handle);

Reviewer: ```suggestion
// Create a SharedMap dynamically at runtime
const newSharedMap = fluidContainer.create(SharedMap);

// BAD: This call won't work; you must store the handle, not the SharedMap itself.
// initialMap.set("newSharedMapKey", newSharedMap);

// GOOD: This call correctly stores the DDS via its handle.
initialMap.set("newSharedMapKey", newSharedMap.handle);
```

```

> Now, anytime you are looking to load the `newSharedMap` back at a later time, all you have to do is load it back from the stored handle.

Reviewer: ```suggestion
To load the `newSharedMap` at a later time, you first retrieve the handle and then retrieve the object from the handle.
```

```javascript
const newSharedMapHandle = initialMap.get("newSharedMapKey");
const newSharedMap = await newSharedMapHandle.get();
```

> {{< callout note >}}

Reviewer: ```suggestion
{{< callout tip >}}
```

> Loading any DDS from its handle is an asynchronous operation. Whereas objects defined in `initialObjects` will be available synchronously once the container is provided, you will need to use an `await` or chain Promises on the `handle.get()` call.

Reviewer: ```suggestion
Loading any DDS from its handle is an asynchronous operation. You will need to use an `await` or chain Promises on the `handle.get()` call to retrieve nested DDSes.
```

{{< /callout >}}

#### Nested shared objects example

> Let's take another look at the example from the **Storing objects** section above to see how it can be updated to use nested `SharedMaps`. This will introduce some data hierarchy and avoid having a flat map that has both the tasks and their fields themselves on the same level. To do so, let's start from the earlier model but with two tasks now.

Reviewer: ```suggestion
You can further extend the example from the [Storing objects]({{< relref "#storing-objects" >}}) section above to see how it can be updated to use nested `SharedMaps`. This will introduce a hierarchy to the data to make it easier to work with. To do so, consider the earlier data model but with two tasks.
```

```json
{
    "task1": {
        "title": "Awesome Task",
        "description": "Doing the most awesome things",
        "assignedTo": {
            "name": "Joe Schmo",
            "email": "joeschmo@email.com",
            "address": "1234 Fluid Way"
        }
    },
    "task2": {
        "title": "Even More Awesome Task",
        "description": "Doing even more awesome things",
        "assignedTo": {
            "name": "Jane Doe",
            "email": "janedoe@email.com",
            "address": "5678 Framework Street"
        }
    }
}
```

> Here, you can have each task itself be its own `SharedMap` and have a parent `SharedMap` hold all of the handles to each task.
> 
> Then, the initial map would look like:
> ```json
> {
>     "task1": task1MapHandle,
>     "task2": task2MapHandle
> }
> ```
> And the `task1` map would have:
> ```json
> {
>     "title": "Awesome Task",
>     "description": "Doing the most awesome things",
>     "assignedTo": {
>         "name": "Joe Schmo",
>         "email": "joeschmo@email.com",
>         "address": "1234 Fluid Way"
>     }
> }

Reviewer: ```suggestion

To break this apart, you can have each task itself be its own `SharedMap` and have a parent `SharedMap` hold all of the handles to each task. Then the initial map would look like:

```json
{
    "task1": task1MapHandle,
    "task2": task2MapHandle
}
```

And the `task1` map would look like:

```json
{
    "title": "Awesome Task",
    "description": "Doing the most awesome things",
    "assignedTo": {
        "name": "Joe Schmo",
        "email": "joeschmo@email.com",
        "address": "1234 Fluid Way"
    }
}
```

```

> And the `task2` map would have:
> ```json

Reviewer: ```suggestion

And the `task2` map would look like:

```json
```

{
    "title": "Even More Awesome Task",
    "description": "Doing even more awesome things",
    "assignedTo": {
        "name": "Jane Doe",
        "email": "janedoe@email.com",
        "address": "5678 Framework Street"
    }
}
```

> With this nested map structure, you are able to both ensure that each field that will be collaboratively edited is stored separately and that you have a hierarchy in how you store the data in your `SharedMap` that reflects the app's data model.
> 
> Whenever a new task is created, you can call `container.create` to get a new `SharedMap` instance and store its handle into the `initialMap` that is provided as an `initialObject`. Since each additional task map would need its own unique key, you can use a random string generator, such as [uuid](https://www.npmjs.com/package/uuid), to create a new one each time and avoid any clashes.
> 

Reviewer: ```suggestion
With this nested map structure, you can both ensure that each field that will be simultaneously edited is stored separately and that you have a hierarchy in how you store the data in your `SharedMap` that reflects the app's data model.

Whenever a new task is created, you can call `container.create` to create a new `SharedMap` instance and store its handle in the `initialMap` that is provided as an `initialObject`. Since each additional task map will need its own unique key, you can use a random string generator, such as [uuid](https://www.npmjs.com/package/uuid), to create a new one each time and avoid ID conflicts.

```

```javascript
const schema = {
    name: "example-container",
    initialObjects: {
        initalMap: SharedMap,
    },
    dynamicObjectTypes: [SharedMap]
}

const { fluidContainer, containerServices } = await client.getContainer(/*service config*/, schema);

const initialMap = fluidContainer.initialObjects.initialMap;

const newTask = fluidContainer.create(SharedMap);
const newTaskId = uuid();

initialMap.set(newTaskId, newTask.handle);
```

> At this point, you can use `intialMap.keys()` to see all of the various task IDs that are available or `initialMap.values()` to return all of the task handles.

Reviewer: ```suggestion
At this point, you can use `intialMap.keys()` to retrieve the IDs of all the tasks or `initialMap.values()` to return the handles to the SharedMaps for each task..
```

> Say you wanted to fetch back a specific task, with ID `task123` and allow the user to edit its description.

Reviewer: ```suggestion
For example, if you wanted to fetch task with ID `task123` and allow the user to edit its description, you would use code like this:
```

```javascript
const taskHandle = initialMap.get("task123");
const task = await taskHandle.get();
task.set("description", editedDescription)
```

> Since each task is being in a separate map and all of the fields within the task object are being stored in their own unique keys, you can now have both a hiearchical structure in your data and avoid any clashes in edits between multiple users.

Reviewer: ```suggestion
Since each task is stored in a separate SharedMap and all of the fields within the task object are being stored in their own unique keys, your data now has a hierarchical structure that reflects the app's data model while individual tasks' properties can be edited independently.

```
